src/code/z_bgcheck.c

```diff
@@ -1,461 +1,4293 @@
+        if ((checkHeight + dy) < 5.0f) {
+            result = func_8003D7F0(colCtx, xpFlags, COLPOLY_IGNORE_NONE, posPrev, posNext, &posIntersect, &poly, &bgId,
+                                   actor, 1.0f, 0x1B); // bccFlags skip ceiling
+            if (result != false) {
+                ny = COLPOLY_GET_NORM(poly->norm.y);
+                // if poly is floor, push result underneath the floor
+                if (0.5f < ny) {
+                    posResult->x = posIntersect.x;
+                    if (1.0f < checkHeight) {
+                        posResult->y = posIntersect.y - 1.0f;
+                    } else {
+                        posResult->y = posIntersect.y - checkHeight;
+                    }
+                    posResult->z = posIntersect.z;
+                }
+                // poly is wall
+                else {
+                    nx = COLPOLY_GET_NORM(poly->norm.x);
+                    nz = COLPOLY_GET_NORM(poly->norm.z);
+                    posResult->x = radius * nx + posIntersect.x;
+                    posResult->y = radius * ny + posIntersect.y;
+                    posResult->z = radius * nz + posIntersect.z;
+                }
+                *outPoly = poly;
+                *outBgId = bgId;
+            }
+        } else {
+            // if the radius is less than the distance travelled on the xz plane, also test for floor collisions
+            bccFlags =
+                SQ(radius) < (SQ(dx) + SQ(dz)) ? 0x1B : 0x19; // bccFlags skip ceiling : bccFlags skip floor and ceiling
+
+            // perform a straight line test to see if a line at posNext.y + checkHeight from posPrev.xz to posNext.xz
+            // passes through any wall and possibly floor polys
+            checkLineNext = *posNext;
+            checkLineNext.y += checkHeight;
+            checkLinePrev = *posPrev;
+            checkLinePrev.y = checkLineNext.y;
+            result = func_8003D7F0(colCtx, xpFlags, COLPOLY_IGNORE_NONE, &checkLinePrev, &checkLineNext, &posIntersect,
+                                   &poly, &bgId, actor, 1.0f, bccFlags);
+
+            if (result != false) {
+                nx2 = COLPOLY_GET_NORM(poly->norm.x);
+                nz2 = COLPOLY_GET_NORM(poly->norm.z);
+                n2XZDist = sqrtf(SQ(nx2) + SQ(nz2));
+
+                // if poly is not a "flat" floor or "flat" ceiling
+                if (!IS_ZERO(n2XZDist)) {
+                    // normalize nx,nz and multiply each by the radius to go back to the other side of the wall
+                    f32temp = 1.0f / n2XZDist;
+                    temp_f0 = radius * f32temp;
+                    posResult->x = temp_f0 * nx2 + posIntersect.x;
+                    posResult->z = temp_f0 * nz2 + posIntersect.z;
+                    *outPoly = poly;
+                    *outBgId = bgId;
+                    result = true;
+                }
+            }
+        }
+    }
+    dynaPos = *posResult;
```

Comment:
```suggestion
    sphCenter = *posResult;
```
since it's also used for static collision